### ISSUE TITLE
From shots to states and expectations

### ISSUE THREAD

**Issue by scarrazza** [Quantum-TII/qibo#264]
We would to have the possibility to convert measurement shots to final state functions and/or to hamiltonian expectation.

Some useful techniques are:
- Quantum tomography
- Quantum Phase Estimation

**Comment by AdrianPerezSalinas** [comment]
Hi @scarrazza , I have been thinking about this problem and have two ideas.

Quantum tomography: if one makes a change of basis in the circuit (i. e., a layer of Hadamard gates) right before the measurements, the output in the computational basis can be written as a linear system in cos(phi_i - phi_j), where phi_i is the complex phase of entry i, and assuming phi_0 = 0. In order to extract the phases, we should perform a classical post_processing allowing us to solve this linear system. This procedure might be more efficient than standard quantum tomography, and thus more useful.

QPE: in case where the exponential of a Hamiltonian can be efficiently implemented in a quantum circuit, we could estimate the overlap between the circuit state and the ground state of that Hamiltonian. It would not serve for computing a expected value, but could be useful for optimizing. 

I have to write both processes carefully, and when I have it, I will send it to you.

**Comment by joseignaciolatorre** [comment]
I do not think there is any free lunch here.
It is a counting of degrees of freedom.
The density matrix has 4^n d.o.f., thus you need 4^n measurements,
that is O_i={I,sigma_x,sigma_y,simga_z} per site, i=1,...,n .

The advantage can be obtained if not all information is needed.
Then, agreed, there might be more efficient procedures.

**Comment by AdrianPerezSalinas** [comment]
Yes, I agree with that, it is a general result of quantum tomography. However, I still think there is something missing here, at least in the case where the density matrix is not a proper density matrix, but only one quantum state. The modulus of all coefficients involve all possible measurements with Z and I, that is a total amount of 2^n measurements. However, in practice it is done by just measuring the output of all qubits in the computational basis (assuming many shots). From these results, it is possible to infer the value of all measurements involving Z and I. 

I have the suspect that something kind of similar can be done with the phases, so that the wavefunction of a quantum circuit can be obtained with few different measurements, although many shots
